In [1]:
import pandas as pd, numpy as np, feather

In [2]:
# Read the data

tasers = pd.read_excel("data/2015 Reported Taser Data.xlsx",
                      header=[0,1])

In [3]:
# Look at the data

tasers.head(5)

Law Enforcement Agency Number of Incident Reports Incident Case Number  \
                               Unnamed: 0_level_1   Unnamed: 1_level_1   
Hartford                                        1             2015-055   
Naugatuck                                       1                15-19   
Norwalk                                         1             15-00010   
CSP (State Police)                              1           1500006763   
Hartford                                        1            2015-0379   

Law Enforcement Agency   Date of Incident   Time of Incident  \
                       Unnamed: 2_level_1 Unnamed: 3_level_1   
Hartford                       2015-01-01           15:13:00   
Naugatuck                      2015-01-01           03:48:00   
Norwalk                        2015-01-01           00:32:00   
CSP (State Police)             2015-01-04           11:41:00   
Hartford                       2015-01-05           12:52:00   

Law Enforcement Agency                Sex               Race  \
                       Unnamed: 4_level_1 Unnamed: 5_level_1   
Hartford                             Male              Black   
Naugatuck                            Male              Black   
Norwalk                              Male              White   
CSP (State Police)                   Male              White   
Hartford                             Male              Black   

Law Enforcement Agency           Hispanic             Height  \
                       Unnamed: 6_level_1 Unnamed: 7_level_1   
Hartford                            False              5'10"   
Naugatuck                           False               5'8"   
Norwalk                             False              5'10"   
CSP (State Police)                  False              5'10"   
Hartford                            False                 6'   

Law Enforcement Agency             Weight                Age  \
                       Unnamed: 8_level_1 Unnamed: 9_level_1   
Hartford                              220                 34   
Naugatuck                             150                 26   
Norwalk                               175                 22   
CSP (State Police)                    150                 35   
Hartford                              153                 62   

Law Enforcement Agency                 ...                  \
                                       ...                   
Hartford                               ...                   
Naugatuck                              ...                   
Norwalk                                ...                   
CSP (State Police)                     ...                   
Hartford                               ...                   

Law Enforcement Agency Subject Resistance Resulting in Application of Force  \
                                                     Threaten Use of Weapon   
Hartford                                                              NaN     
Naugatuck                                                             NaN     
Norwalk                                                               NaN     
CSP (State Police)                                                    NaN     
Hartford                                                              NaN     

Law Enforcement Agency                                             \
                       Fleeing Unarmed Assault Armed with Firearm   
Hartford                   NaN             NaN                NaN   
Naugatuck                  NaN             NaN                NaN   
Norwalk                    NaN             NaN                NaN   
CSP (State Police)         NaN               1                NaN   
Hartford                   NaN             NaN                NaN   

Law Enforcement Agency                                                      \
                       Armed with Edged Weapon Armed with Blunt Instrument   
Hartford                                   NaN          

In [4]:
# Count the number of rows in the dataset to check that future
# operations work as expected

max(tasers.count())

610

In [5]:
for i in tasers.columns:
    print i

(u'Number of Incident Reports', 'Unnamed: 0_level_1')
(u'Incident Case Number', 'Unnamed: 1_level_1')
(u'Date of Incident', 'Unnamed: 2_level_1')
(u'Time of Incident', 'Unnamed: 3_level_1')
(u'Sex', 'Unnamed: 4_level_1')
(u'Race', 'Unnamed: 5_level_1')
(u'Hispanic', 'Unnamed: 6_level_1')
(u'Height', 'Unnamed: 7_level_1')
(u'Weight', 'Unnamed: 8_level_1')
(u'Age', 'Unnamed: 9_level_1')
(u'Deployment Type I', 'Unnamed: 10_level_1')
(u'Deployment Type II', 'Unnamed: 11_level_1')
(u'Time/Length', u'Displays or Arc')
(u'Time/Length', u'Drive Stun Application')
(u'Time/Length', u'Activation After Probe Contact')
(u'Number of Deployments', u'Activation After Probe Contact')
(u'Warning to Subject', u'Activation After Probe Contact')
(u"Subject's Injuries (If Applicable)", u'Subject Injured')
(u"Subject's Injuries (If Applicable)", u'Not Injured')
(u"Subject's Injuries (If Applicable)", u'Bruises')
(u"Subject's Injuries (If Applicable)", u'Abrasions')
(u"Subject's Injuries (If Applicable)", u'B

In [6]:
# There's information on the subjects' height and weight.
# Do bigger people get tasered more often?

tasers.reset_index()["Weight"]["Unnamed: 8_level_1"].value_counts()\
.sort_values(ascending=False)

Unknown    51
180        50
200        44
160        42
170        39
150        36
175        30
190        26
165        22
250        19
140        18
220        18
230        18
210        14
215        12
155        11
130        11
300        11
185        10
240        10
145         9
195         8
260         8
225         7
205         7
120         6
270         5
290         4
100         3
275         3
           ..
156         1
158         1
207         1
172         1
183         1
163         1
153         1
191         1
171         1
189         1
173         1
174         1
198         1
129         1
170-180     1
245         1
115         1
370         1
229         1
154         1
212         1
226         1
144         1
152         1
146         1
122         1
148         1
305         1
177         1
128         1
Name: Unnamed: 8_level_1, dtype: int64

In [7]:
# Let's look at the median and mean weights of people who are tasered
# Let's get rid of non_numeric values like ("170-180")

pd.to_numeric(tasers.reset_index()["Weight"]["Unnamed: 8_level_1"],
              errors="coerce").mean()

188.83273381294964

In [8]:
pd.to_numeric(tasers.reset_index()["Weight"]["Unnamed: 8_level_1"],
              errors="coerce").median()

180.0

### Weight: 180-190 lbs.

The average weight of subjects involved in taser incidents is 180 to 190 pounds. The median weight is 180.

In [9]:
# Height will be a little tricker. We have to convert feet'inches" to inches

In [10]:
tasers["Height"]

Unnamed: 7_level_1
Hartford                               5'10"
Naugatuck                               5'8"
Norwalk                                5'10"
CSP (State Police)                     5'10"
Hartford                                  6'
Cheshire                                5'8"
East Windsor                            6'2"
Danbury                                5'10"
New London                              5'2"
Shelton                                5'11"
New Milford                             6'3"
New Haven                               5'7"
Trumbull                                  6'
Norwalk                                 5'2"
New London                             5'10"
CSP (State Police)                      5'9"
New Britain                             6'0"
Enfield                                 5'8"
New London                              5'8"
Norwalk                                 6'2"
Norwalk                                 5'4"
Meriden                                5'10"
Suffield                                6'2"
Norwalk                                5'10"
Vernon                                 5'11"
Norwalk                                 5'8"
Manchester                              5'7"
Wallingford                             6'4"
Branford                               5'11"
CSP (State Police)                      5'7"
...                                      ...
Meriden                                 6'0"
University of Connecticut               5'8"
CSP (State Police)                     5'10"
Hartford                                5'8"
East Hartford                           6'0"
Hartford                                5'8"
Derby                                   5'9"
Cheshire                                5'6"
Waterbury                               6'0"
New London                              5'9"
Clinton                                5'11"
Glastonbury                             5'9"
Norwalk                                 6'2"
CSP (State Police)                      6'4"
Clinton                                   6'
Winchester                           Unknown
Easton                                  5'6"
Shelton                                 5'5"
Waterbury                               6'5"
Norwalk                                 6'2"
Meriden                                 5'7"
Naugatuck                                 6'
Trumbull                                  6'
Norwalk                                5'10"
Torrington                              5'7"
Stratford                               5'7"
East Haven                              5'7"
Newington                               6'3"
Wilton                               Unknown
West Hartford                           5'8"

[610 rows x 1 columns]

In [11]:
## Let's make a function to convert these to useful heights in inches

def height_to_inches(h):
    print "Converting height '" + str(h) + "'"
    
    h = str(h)
    h = h.strip()
    
    ''' We will say everything before the single quote is feet
    and everything after is inches, and if it doesn't fit this format
    return null '''
    
    if "'" not in h:
        return None

    quote_i = h.index("'")
    feet = h[:quote_i]
    #print int(feet)
    
    ## IF there is no " denoting inches, we're done
    if "\"" not in h:
        return 12 * int(feet)
    
    inches = h[quote_i+1:len(h) - 1]
    #print inches
    
    return int(feet) * 12 + int(inches)

height_to_inches("5'8\"")
#height_to_inches("6'11")

#height_to_inches("Unknown")

Converting height '5'8"'


68

In [12]:
# Test out the height conversion function
height_to_inches("5'0\"") == 60


Converting height '5'0"'


True

In [13]:
height_to_inches("0'0\"") == 0


Converting height '0'0"'


True

In [14]:
height_to_inches("6'0\"") == 72

Converting height '6'0"'


True

In [15]:
height_to_inches("Unknown") == None

Converting height 'Unknown'


True

In [16]:
tasers["Height"]["Unnamed: 7_level_1"].apply(lambda x: height_to_inches(x))

Converting height '5'10"'
Converting height '5'8"'
Converting height '5'10"'
Converting height '5'10"'
Converting height '6''
Converting height '5'8"'
Converting height '6'2"'
Converting height '5'10"'
Converting height '5'2"'
Converting height '5'11"'
Converting height '6'3"'
Converting height '5'7"'
Converting height '6''
Converting height '5'2"'
Converting height '5'10"'
Converting height '5'9"'
Converting height '6'0"'
Converting height '5'8"'
Converting height '5'8"'
Converting height '6'2"'
Converting height '5'4"'
Converting height '5'10"'
Converting height '6'2"'
Converting height '5'10"'
Converting height '5'11"'
Converting height '5'8"'
Converting height '5'7"'
Converting height '6'4"'
Converting height '5'11"'
Converting height '5'7"'
Converting height '5'09"'
Converting height '5'6"'
Converting height '6'1"'
Converting height '5'11"'
Converting height '5'7"'
Converting height '5'7"'
Converting height '6'0"'
Converting height '5'3"'
Converting height '6''
Converting height '

Hartford                     70
Naugatuck                    68
Norwalk                      70
CSP (State Police)           70
Hartford                     72
Cheshire                     68
East Windsor                 74
Danbury                      70
New London                   62
Shelton                      71
New Milford                  75
New Haven                    67
Trumbull                     72
Norwalk                      62
New London                   70
CSP (State Police)           69
New Britain                  72
Enfield                      68
New London                   68
Norwalk                      74
Norwalk                      64
Meriden                      70
Suffield                     74
Norwalk                      70
Vernon                       71
Norwalk                      68
Manchester                   67
Wallingford                  76
Branford                     71
CSP (State Police)           67
                             ..
Meriden 

In [17]:
## Finally, let's convert the heights save them to a new "calculated" column

# tasers["Calculated"]["Height"] = tasers["Height"]["Unnamed: 7_level_1"].apply(lambda x: height_to_inches(x))

# This multi-index thing is messing me up... 

tasers["Calculated", "Height"] = tasers["Height"]["Unnamed: 7_level_1"].apply(lambda x: height_to_inches(x))

Converting height '5'10"'
Converting height '5'8"'
Converting height '5'10"'
Converting height '5'10"'
Converting height '6''
Converting height '5'8"'
Converting height '6'2"'
Converting height '5'10"'
Converting height '5'2"'
Converting height '5'11"'
Converting height '6'3"'
Converting height '5'7"'
Converting height '6''
Converting height '5'2"'
Converting height '5'10"'
Converting height '5'9"'
Converting height '6'0"'
Converting height '5'8"'
Converting height '5'8"'
Converting height '6'2"'
Converting height '5'4"'
Converting height '5'10"'
Converting height '6'2"'
Converting height '5'10"'
Converting height '5'11"'
Converting height '5'8"'
Converting height '5'7"'
Converting height '6'4"'
Converting height '5'11"'
Converting height '5'7"'
Converting height '5'09"'
Converting height '5'6"'
Converting height '6'1"'
Converting height '5'11"'
Converting height '5'7"'
Converting height '5'7"'
Converting height '6'0"'
Converting height '5'3"'
Converting height '6''
Converting height '

In [18]:
tasers["Calculated"]["Height"].mean()

69.737302977232929

In [19]:
# Median height was six feet tall
tasers["Calculated"]["Height"].median()

70.0

In [20]:
tasers["Calculated"]["Height"].describe()

count    571.000000
mean      69.737303
std        3.340500
min       59.000000
25%       68.000000
50%       70.000000
75%       72.000000
max       87.000000
Name: Height, dtype: float64

In [21]:
print "The tallest person was " + str(tasers["Calculated"]["Height"].max()/12) + " feet tall"

The tallest person was 7.25 feet tall


In [22]:
print "The shortest person was " + str(tasers["Calculated"]["Height"].min()/12) + " feet tall"

The shortest person was 4.91666666667 feet tall


## Height, around six feet tall

The median and mean height of people who were subject to stun gun use was around six feet tall.

In [23]:
# Let's take a look at the different outcomes by height and weight
# Did smaller people fall into the category of being warned but not stunned more often?



In [24]:
## Deployment type I and Deployment Type II seem to be the best column to determine 
## if people were warned or tasered. The format is definitely not ideal, having a fixed
## Number of columns for each specific taser deployment, so there can only be a record
## of at maximum two deployments. Not sure what happens if more than two officers
## fire tasers. How would that be reflected in this data set?

#(u'Deployment Type I', 'Unnamed: 10_level_1')
#(u'Deployment Type II', 'Unnamed: 11_level_1')

# Here's what the different codes mean, according to the data dictionary:

# This field indicates the second type of taser deployment if there was more than one 
#(either by multiple officers or different methods by the same officer). 

# The first three values indicate that the subject was not stunned

# U=Un-Holster, 
# L=Laser Sight, 
# W=Warning Arc, 

# The following indicate an actual firing of the taser. 

# C=Cartridge, 
# D=Drive Stun

In [25]:
# Primary firings
tasers["Deployment Type I"]["Unnamed: 10_level_1"].value_counts()

C    326
L    175
D     93
U      9
W      7
Name: Unnamed: 10_level_1, dtype: int64

In [26]:
# Secondary firings
tasers["Deployment Type II"]["Unnamed: 11_level_1"].value_counts()

D     55
L      9
C      3
Name: Unnamed: 11_level_1, dtype: int64

In [27]:
# How many records have non-null values in these categories?
tasers["Deployment Type I"]["Unnamed: 10_level_1"].count()

610

In [28]:
tasers["Deployment Type II"]["Unnamed: 11_level_1"].count()

67

In [29]:
## So that's good. Every record has at least a primary deployment type

In [30]:
## C + D is the number of people actually fired at
tasers[(tasers["Deployment Type I"]["Unnamed: 10_level_1"] == "C") 
             | (tasers["Deployment Type I"]["Unnamed: 10_level_1"] == "D")]\
       ["Deployment Type I"]["Unnamed: 10_level_1"].count()

419

In [31]:
tasers[(tasers["Deployment Type II"]["Unnamed: 11_level_1"] == "C") 
             | (tasers["Deployment Type II"]["Unnamed: 11_level_1"] == "D")]\
       ["Deployment Type I"]["Unnamed: 10_level_1"].count()

58

In [32]:
tasers_weight_outcomes = pd.to_numeric(tasers["Calculated"]["Height"].to_frame(), errors='coerce')
tasers_weight_outcomes["weight"] = pd.to_numeric(tasers["Weight"]["Unnamed: 8_level_1"], errors='coerce')
tasers_weight_outcomes["deployment_primary"] = tasers["Deployment Type I"]["Unnamed: 10_level_1"]
tasers_weight_outcomes["deployment_secondary"] = tasers["Deployment Type II"]["Unnamed: 11_level_1"]

tasers_weight_outcomes.columns=["height","weight","deployment_primary","deployment_secondary"]


In [33]:
tasers_weight_outcomes.groupby("deployment_primary").agg({'height':np.median,'weight':np.median}).reset_index()

deployment_primary  weight  height
0                  C   180.0      70
1                  D   180.0      70
2                  L   180.0      70
3                  U   167.5      71
4                  W   160.0      71

In [34]:
tasers_weight_outcomes.groupby("deployment_primary").agg({'height':np.mean,'weight':np.mean}).reset_index()

deployment_primary      weight     height
0                  C  192.837209  69.693069
1                  D  189.602410  69.806818
2                  L  181.831250  69.713415
3                  U  177.125000  70.333333
4                  W  175.000000  70.571429

In [35]:
# U=Un-Holster, 
# L=Laser Sight, 
# W=Warning Arc, 

# C=Cartridge, 
# D=Drive Stun

In [36]:
tasers_weight_outcomes.groupby("deployment_secondary").agg({'height':np.median,'weight':np.median}).reset_index()

deployment_secondary  weight  height
0                    C   200.0      67
1                    D   192.5      69
2                   L    180.0      71

In [37]:
# What about secondary taser deployments? 

tasers_weight_outcomes.groupby("deployment_secondary").agg({'height':np.mean,'weight':np.mean}).reset_index()

deployment_secondary      weight     height
0                    C  206.666667  69.333333
1                    D  209.846154  69.716981
2                   L   198.888889  71.111111

In [38]:
# Just a reminder here, the totals for each type of outcome. Far fewer people were tased a second time. 

tasers_weight_outcomes["deployment_primary"].value_counts()

C    326
L    175
D     93
U      9
W      7
Name: deployment_primary, dtype: int64

In [39]:
tasers_weight_outcomes["deployment_secondary"].value_counts()

D     55
L      9
C      3
Name: deployment_secondary, dtype: int64

## Bigger people tasered more

From the above calulations it appears the people who were actually fired upon, either with a cartridge or "drive stun" were a little bit heavier on average, but about the same height. 

A little more than 10 percent of subjects were subject to a secondary taser deployment (such as from a second officer). Those people tended to be heavier than the people who were only fired upon once.

****

# Topic 2: Time of day

In [40]:
tasers["Time of Incident"]["Unnamed: 3_level_1"]

Hartford                     15:13:00
Naugatuck                    03:48:00
Norwalk                      00:32:00
CSP (State Police)           11:41:00
Hartford                     12:52:00
Cheshire                     18:07:00
East Windsor                 23:54:00
Danbury                      09:00:00
New London                   10:18:00
Shelton                      11:30:00
New Milford                  13:51:00
New Haven                    01:53:00
Trumbull                     01:09:00
Norwalk                      06:40:00
New London                   03:00:00
CSP (State Police)           18:20:00
New Britain                  01:30:00
Enfield                      12:50:00
New London                   00:36:00
Norwalk                      22:56:00
Norwalk                      22:56:00
Meriden                      18:46:00
Suffield                     16:15:00
Norwalk                      11:22:00
Vernon                       10:21:00
Norwalk                      01:29:00
Manchester  

In [41]:
times = tasers["Time of Incident"]["Unnamed: 3_level_1"]

In [42]:
hours = pd.to_datetime(times,format="%H:%M:%S",errors="coerce").dt.hour

In [43]:
hours.count()

608

### Number of incidents by hour of the day

In [44]:
#hours.value_counts()
hours.value_counts().sort_index()

0     26
1     42
2     25
3     21
4      8
5      5
6      6
7      7
8     14
9     15
10    20
11    23
12    24
13    21
14    24
15    28
16    28
17    30
18    35
19    45
20    40
21    38
22    45
23    38
Name: Unnamed: 3_level_1, dtype: int64

### Time topic: dud?

Doesn't look too exciting. Very few people getting tasered at 5 a.m. 

Maybe the only interesting time-of-day angle is how steep they drop off between 4 and 7 a.m., but then pick up again. 

Other than that, there's not too much here.

## Thanks for watching

I'll upload this analysis to TrendCT's data repository at github.com/trendct/

## Addendum, exporting to feather file

For friendly sharing with R. #SharingIsCaring

In [50]:
feather.write_dataframe(tasers, "tasers.feather")


FeatherError: Invalid: unhandled python type, index 0

Law Enforcement Agency    Number of Incident Reports Incident Case Number  \
                                  Unnamed: 0_level_1   Unnamed: 1_level_1   
Hartford                                           1             2015-055   
Naugatuck                                          1                15-19   
Norwalk                                            1             15-00010   
CSP (State Police)                                 1           1500006763   
Hartford                                           1            2015-0379   
Cheshire                                           1             2015-341   
East Windsor                                       1             15-16-AR   
Danbury                                            1           1500001035   
New London                                         1            15-000071   
Shelton                                            1               15-468   
New Milford                                        1           15-1449-AR   
New Haven                                          1              15-1473   
Trumbull                                           1           1500000450   
Norwalk                                            1             15-01796   
New London                                         1            15-000003   
CSP (State Police)                                 1           1500031332   
New Britain                                        1              15-1393   
Enfield                                            1          C2015-03070   
New London                                         1            15-000198   
Norwalk                                            2           15-02887-2   
Norwalk                                            1           15-02887-1   
Meriden                                            1              15-0534   
Suffield                                           1              15-9-AR   
Norwalk                                            1             15-03390   
Vernon                                             1              15-1185   
Norwalk                                            1             15-04010   
Manchester                                         1           M201500936   
Wallingford                                        1            2015-1829   
Branford                                           1              15-2484   
CSP (State Police)                                 1           1500057386   
...                                              ...                  ...   
Meriden                                            1             15-10214   
University of Connecticut                          1           1500054364   
CSP (State Police)                                 1           1500717147   
Hartford                                           1           2015-38437   
East Hartford                                      1             15-32876   
Hartford                                           1            2015-3853   
Derby                                              1             15-13449   
Cheshire                                           1           2015-25870   
Waterbury                                          1           2015-68501   
New London                                         1            15-004232   
Clinton                                            1          15-29615-AR   
Glastonbury                                        1            15-016161   
Norwalk                                            1             15-56119   
CSP (State Police)                                 1           1500733331   
Clinton                                            1             15-30010   
Winchester                                         1              15-8046   
Easton                                             1              15-7993   
Shelton                                            2             15-19444   
Waterbury                                          1             15-70562

In [61]:
# Export only the columns I transformed, since feather doesn't play well with all python objects
# https://github.com/wesm/feather/issues/121
# https://github.com/wesm/feather/blob/master/README.md

feather.write_dataframe(tasers[[("Incident Case Number","Unnamed: 1_level_1"),("Calculated","Height")]],"tasers_height.feather")